bridge-fungible-token/src/main.sw

```diff
@@ -1,69 +1,186 @@
+const DECIMALS = 9u8;
+const LAYER_1_DECIMALS = 18u8;
+
+////////////////////////////////////////
+// Storage declarations
+////////////////////////////////////////
 storage {
-    counter: u64 = 0,
-    data1: ContractId = ~ContractId::from(ZERO_B256),
-    data2: u64 = 0,
-    data3: b256 = ZERO_B256,
-    data4: Address = ~Address::from(ZERO_B256),
+    refund_amounts: StorageMap<(b256, EvmAddress), U256> = StorageMap {},
 }
 
+////////////////////////////////////////
+// ABI Implementations
+////////////////////////////////////////
 // Implement the process_message function required to be a message receiver
 impl MessageReceiver for Contract {
     #[storage(read, write)]
     fn process_message(msg_idx: u8) {
-        storage.counter = storage.counter + 1;
+        let input_sender = input_message_sender(1);
 
-        // Parse the message data
-        let data_length = input_message_data_length(msg_idx);
-        if (data_length >= 32) {
-            let contract_id: b256 = input_message_data(msg_idx, 0);
-            storage.data1 = ~ContractId::from(contract_id);
-        }
-        if (data_length >= 32 + 8) {
-            let num: u64 = input_message_data(msg_idx, 32);
-            storage.data2 = num;
-        }
-        if (data_length >= 32 + 8 + 32) {
-            let big_num: b256 = input_message_data(msg_idx, 32 + 8);
-            storage.data3 = big_num;
-        }
-        if (data_length >= 32 + 8 + 32 + 32) {
-            let address: b256 = input_message_data(msg_idx, 32 + 8 + 32);
-            storage.data4 = ~Address::from(address);
+        require(input_sender.value == LAYER_1_ERC20_GATEWAY, BridgeFungibleTokenError::UnauthorizedUser);
+
+        let message_data = parse_message_data(msg_idx);
+
+        // @review verify `l1_asset` matches hardcoded LAYER_1_TOKEN value
+        require(message_data.l1_asset == ~EvmAddress::from(LAYER_1_TOKEN), BridgeFungibleTokenError::IncorrectAssetDeposited);
+
+        // @review verify `from` matches hardcoded PREDICATE_ROOT value
+        require(message_data.from == ~Address::from(PREDICATE_ROOT), BridgeFungibleTokenError::UnauthorizedUser);
+
+        // @review verify `to` matches hardcoded value TO value
+        require(message_data.to == ~Address::from(TO), BridgeFungibleTokenError::UnauthorizedUser);
+
+        // check that value sent as uint256 can fit inside a u64, else register a refund.
+        let decomposed = decompose(message_data.amount);
+        let amount = ~U256::from(decomposed.0, decomposed.1, decomposed.2, decomposed.3);
+        let l1_amount_opt = amount.as_u64();
+        match l1_amount_opt {
+            Result::Err(e) => {
+                storage.refund_amounts.insert((
+                    message_data.to.value,
+                    message_data.l1_asset,
+                ), amount);
+                // @review emit event (i.e: `DepositFailedEvent`) here to allow the refund process to be initiated?
+            },
+            Result::Ok(v) => {
```

Comment:
The value needs to be converted from the Ethereum side decimals into the Fuel side decimals. This is also an area that could result in a refund (value too large to fit in u64 under new decimals or too small to fit in new decimals)